platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/sources/CustomGeometrySource.java

```diff
@@ -143,19 +145,49 @@ private void cancelTile(int z, int x, int y) {
     }
   }
 
+  @AnyThread
+  private ExecutorService obtainExecutorService() {
+    if (executorService == null || executorService.isShutdown()) {
```

Comment:
Is there a chance for a race condition here.  Client calls onStop, then the C++ code calls fetch (or while it is calling fetch).  the fetchTile call creates a new ExecutorService, which then gets leaked.

Might be better to have the finalize method shutdown and the native cleanup.  Then fetchTile will need to handle RejectedExecutionException if a request is made as the source is getting cleaned up.

Reply:
thanks for :eyes: , agreed on the possible race condition, will patch that up or take a different approach. The suggestion of finalize though would result in the same situation as before, stacking up threads with orientation changes/multiple activities.